jpc/src/tag_tree.rs

```diff
@@ -0,0 +1,193 @@
+use log::info;
+
+/// A tag tree represents a 2d-array of natural numbers.
+///
+/// B.10.2 A tag tree is a way of representing a two-dimensional array of non-negative integers in
+/// a hierarchical way. It successively creates reduced resolution levels of this two-dimensional
+/// array, forming a tree. At every node of this tree the minimum integer of the (up to four) nodes
+/// below it is recorded. Figure B.12 shows an example of this representation. The notation, qi(m,
+/// n), is the value at the node that is mth from the left and nth from the top, at the ith level.
+/// Level 0 is the lowest level of the tag tree; it contains the top node.
+pub struct TagTree {}
+
+/// A decoder from tag tree bits to numbers in the 2d-array.
+///
+/// TagTreeDecoder takes in bits and returns values from the represented 2d-array. Only positive
+/// integers will ever be produced.
+#[derive(Debug)]
+pub struct TagTreeDecoder {
+    max_depth: usize,
+    cur_depth: usize,
+    cur_value: u8,
+    /// levels is a Vec containing data (width, items) for each level of the tree.
+    levels: Vec<(usize, Vec<u8>)>, // Todo, what should be here?
+}
+
+impl TagTreeDecoder {
+    pub fn new(width: usize, height: usize) -> Self {
+        let mut mw = width;
+        let mut mh = height;
+        let mut max_depth = 0;
+        let mut levels = Vec::new();
+        // Determine max depth by dividing out groups of 4
+        while mw > 1 || mh > 1 {
+            let w = mw.max(1);
+            let size: usize = w * mh.max(1);
+            levels.push((w, Vec::with_capacity(size)));
+            println!("added vec of size {size}");
+            max_depth += 1;
+            mw = mw.div_ceil(2);
+            mh = mh.div_ceil(2);
+        }
+        levels.push((1, Vec::with_capacity(1)));
+        levels.reverse(); // reverse in place so level 0 is at index 0
+        println!("Need a depth of {max_depth} to represent tag tree");
+
+        assert_eq!(max_depth + 1, levels.len());
+
+        Self {
+            max_depth,
+            cur_depth: 0,
+            cur_value: 0,
+            levels,
+        }
+    }
+
+    fn cur_offset(&self) -> usize {
+        match self.levels.get(self.cur_depth) {
+            None => panic!("Not deep enough to know current location."),
+            Some((cw, clvl)) => clvl.len(),
+        }
+    }
+
+    // TODO fix push_bit type signature for better return type
+    pub fn push_bit(&mut self, b: u8) -> Option<u8> {
+        if b == 0 {
```

Comment:
T.800 B.10.2 states "For each contiguous 0 bit in the tag tree the current value is incremented by one" but checking if the current byte is zero does not guarantee the previous one is also zero which is how I would interpret this.